File: backend/services/merit_service.py
"""
Merit Service

Business logic layer for Merit operations including validation and processing.
"""

from typing import Optional, List, Dict, Any
from datetime import date
import uuid

from repositories.merit_repository import MeritRepository
from models.merit import Merit


class MeritService:
    """Service for merit business logic"""

    def __init__(self, repository: MeritRepository):
        self.repository = repository

    async def award_merit(
        self,
        school_id: uuid.UUID,
        student_id: uuid.UUID,
        awarded_by_id: uuid.UUID,
        category: str,
        points: int,
        reason: str,
        **kwargs
    ) -> Merit:
        """
        Award a merit to a student with validation

        Args:
            school_id: School ID
            student_id: Student receiving the merit
            awarded_by_id: User awarding the merit
            category: Merit category
            points: Merit points (1-10)
            reason: Reason for awarding
            **kwargs: Additional merit fields

        Returns:
            Created merit

        Raises:
            ValueError: If validation fails
        """
        # Validate category
        valid_categories = ['academic', 'behavior', 'participation', 'leadership', 'attendance', 'other']
        if category not in valid_categories:
            raise ValueError(f"Invalid category. Must be one of: {', '.join(valid_categories)}")

        # Validate points
        if not (1 <= points <= 10):
            raise ValueError("Merit points must be between 1 and 10")

        # Validate reason
        if not reason or len(reason.strip()) < 10:
            raise ValueError("Reason must be at least 10 characters")

        # Validate quarter if provided
        if 'quarter' in kwargs and kwargs['quarter']:
            valid_quarters = ['Q1', 'Q2', 'Q3', 'Q4']
            if kwargs['quarter'] not in valid_quarters:
                raise ValueError(f"Invalid quarter. Must be one of: {', '.join(valid_quarters)}")

        # Create merit
        merit_data = {
            'school_id': school_id,
            'student_id': student_id,
            'awarded_by_id': awarded_by_id,
            'category': category,
            'points': points,
            'reason': reason.strip(),
            'created_by': awarded_by_id,
            **kwargs
        }

        merit = await self.repository.create(merit_data)
        return merit

    async def award_batch_merits(
        self,
        school_id: uuid.UUID,
        student_ids: List[uuid.UUID],
        awarded_by_id: uuid.UUID,
        category: str,
        points: int,
        reason: str,
        **kwargs
    ) -> List[Merit]:
        """
        Award merits to multiple students (class award)

        Args:
            school_id: School ID
            student_ids: List of student IDs
            awarded_by_id: User awarding the merits
            category: Merit category
            points: Merit points (1-10)
            reason: Reason for awarding
            **kwargs: Additional merit fields

        Returns:
            List of created merits

        Raises:
            ValueError: If validation fails
        """
        # Validate at least one student
        if not student_ids or len(student_ids) == 0:
            raise ValueError("At least one student ID is required for batch award")

        # Validate category
        valid_categories = ['academic', 'behavior', 'participation', 'leadership', 'attendance', 'other']
        if category not in valid_categories:
            raise ValueError(f"Invalid category. Must be one of: {', '.join(valid_categories)}")

        # Validate points
        if not (1 <= points <= 10):
            raise ValueError("Merit points must be between 1 and 10")

        # Validate reason
        if not reason or len(reason.strip()) < 10:
            raise ValueError("Reason must be at least 10 characters")

        # Generate batch ID
        batch_id = uuid.uuid4()

        # Create merit data for each student
        merits_data = []
        for student_id in student_ids:
            merit_data = {
                'school_id': school_id,
                'student_id': student_id,
                'awarded_by_id': awarded_by_id,
                'category': category,
                'points': points,
                'reason': reason.strip(),
                'created_by': awarded_by_id,
                **kwargs
            }
            merits_data.append(merit_data)

        # Create batch
        merits = await self.repository.create_batch(merits_data, batch_id)
        return merits

    async def update_merit(
        self,
        merit_id: uuid.UUID,
        updated_by_id: Optional[uuid.UUID] = None,
        **kwargs
    ) -> Merit:
        """
        Update merit with validation

        Args:
            merit_id: Merit ID
            updated_by_id: User updating the merit
            **kwargs: Fields to update

        Returns:
            Updated merit

        Raises:
            ValueError: If validation fails
        """
        merit = await self.repository.get_by_id(merit_id)
        if not merit:
            raise ValueError(f"Merit with ID {merit_id} not found")

        # Validate category if changing
        if 'category' in kwargs:
            valid_categories = ['academic', 'behavior', 'participation', 'leadership', 'attendance', 'other']
            if kwargs['category'] not in valid_categories:
                raise ValueError(f"Invalid category. Must be one of: {', '.join(valid_categories)}")

        # Validate points if changing
        if 'points' in kwargs:
            if not (1 <= kwargs['points'] <= 10):
                raise ValueError("Merit points must be between 1 and 10")

        # Validate reason if changing
        if 'reason' in kwargs:
            if not kwargs['reason'] or len(kwargs['reason'].strip()) < 10:
                raise ValueError("Reason must be at least 10 characters")
            kwargs['reason'] = kwargs['reason'].strip()

        # Validate quarter if changing
        if 'quarter' in kwargs and kwargs['quarter']:
            valid_quarters = ['Q1', 'Q2', 'Q3', 'Q4']
            if kwargs['quarter'] not in valid_quarters:
                raise ValueError(f"Invalid quarter. Must be one of: {', '.join(valid_quarters)}")

        kwargs['updated_by'] = updated_by_id
        updated_merit = await self.repository.update(merit_id, kwargs)
        return updated_merit

    async def revoke_merit(
        self,
        merit_id: uuid.UUID,
        deleted_by_id: Optional[uuid.UUID] = None
    ) -> None:
        """
        Revoke a merit (soft delete)

        Args:
            merit_id: Merit ID
            deleted_by_id: User revoking the merit
        """
        merit = await self.repository.get_by_id(merit_id)
        if not merit:
            raise ValueError(f"Merit with ID {merit_id} not found")

        await self.repository.delete(merit_id)

    async def get_student_summary(self, student_id: uuid.UUID) -> Dict[str, Any]:
        """
        Get comprehensive merit summary for a student

        Args:
            student_id: Student ID

        Returns:
            Dictionary with student merit summary
        """
        summary = await self.repository.get_student_summary(student_id)
        return summary

    async def get_class_summary(
        self,
        class_id: uuid.UUID,
        quarter: Optional[str] = None
    ) -> Dict[str, Any]:
        """
        Get class merit statistics

        Args:
            class_id: Class ID
            quarter: Optional quarter filter

        Returns:
            Dictionary with class statistics
        """
        summary = await self.repository.get_class_summary(class_id, quarter)
        return summary

    async def get_leaderboard(
        self,
        school_id: uuid.UUID,
        grade_level: Optional[int] = None,
        quarter: Optional[str] = None,
        limit: int = 20
    ) -> List[Dict[str, Any]]:
        """
        Get merit leaderboard

        Args:
            school_id: School ID
            grade_level: Optional grade level filter
            quarter: Optional quarter filter
            limit: Number of top students to return

        Returns:
            List of student rankings
        """
        leaderboard = await self.repository.get_leaderboard(
            school_id=school_id,
            grade_level=grade_level,
            quarter=quarter,
            limit=limit
        )
        return leaderboard

    async def get_statistics(
        self,
        school_id: uuid.UUID,
        quarter: Optional[str] = None,
        grade_level: Optional[int] = None
    ) -> Dict[str, Any]:
        """
        Get school-wide merit statistics

        Args:
            school_id: School ID
            quarter: Optional quarter filter
            grade_level: Optional grade level filter

        Returns:
            Dictionary with comprehensive statistics
        """
        stats = await self.repository.get_statistics(
            school_id=school_id,
            quarter=quarter,
            grade_level=grade_level
        )
        return stats

File: frontend/src/services/meritService.ts
/**
 * Merit Service
 * API client for merit operations
 */

import type {
  Merit,
  MeritCreateInput,
  MeritBatchCreateInput,
  MeritUpdateInput,
  MeritListResponse,
  MeritSummary,
  ClassMeritSummary,
  LeaderboardEntry,
  MeritStatistics,
  MeritFilters,
} from '@/types/merit'

const API_BASE_URL = import.meta.env.VITE_API_BASE_URL || 'http://localhost:8000'

class MeritService {
  private baseUrl = `${API_BASE_URL}/api/v1/merits`

  /**
   * Get all merits with optional filters
   */
  async getMerits(
    schoolId: string,
    filters?: MeritFilters
  ): Promise<MeritListResponse> {
    const params = new URLSearchParams({
      school_id: schoolId,
      page: String(filters?.page || 1),
      limit: String(filters?.limit || 50),
    })

    if (filters?.category) {
      params.append('category', filters.category)
    }
    if (filters?.quarter) {
      params.append('quarter', filters.quarter)
    }
    if (filters?.awarded_by_id) {
      params.append('awarded_by_id', filters.awarded_by_id)
    }
    if (filters?.start_date) {
      params.append('start_date', filters.start_date)
    }
    if (filters?.end_date) {
      params.append('end_date', filters.end_date)
    }

    const response = await fetch(`${this.baseUrl}?${params}`)
    if (!response.ok) {
      throw new Error(`Failed to fetch merits: ${response.statusText}`)
    }
    return response.json()
  }

  /**
   * Get merit by ID
   */
  async getMeritById(meritId: string): Promise<Merit> {
    const response = await fetch(`${this.baseUrl}/${meritId}`)
    if (!response.ok) {
      throw new Error(`Failed to fetch merit: ${response.statusText}`)
    }
    return response.json()
  }

  /**
   * Award merit to a student
   */
  async awardMerit(
    meritData: MeritCreateInput,
    awardedById: string
  ): Promise<Merit> {
    const params = new URLSearchParams({ awarded_by_id: awardedById })

    const response = await fetch(`${this.baseUrl}?${params}`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(meritData),
    })

    if (!response.ok) {
      const error = await response.json()
      throw new Error(error.detail || `Failed to award merit: ${response.statusText}`)
    }
    return response.json()
  }

  /**
   * Award merits to multiple students (batch/class award)
   */
  async awardBatchMerits(
    meritData: MeritBatchCreateInput,
    awardedById: string
  ): Promise<Merit[]> {
    const params = new URLSearchParams({ awarded_by_id: awardedById })

    const response = await fetch(`${this.baseUrl}/batch?${params}`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(meritData),
    })

    if (!response.ok) {
      const error = await response.json()
      throw new Error(error.detail || `Failed to award batch merits: ${response.statusText}`)
    }
    return response.json()
  }

  /**
   * Update merit
   */
  async updateMerit(
    meritId: string,
    meritData: MeritUpdateInput,
    updatedById: string
  ): Promise<Merit> {
    const params = new URLSearchParams({ updated_by_id: updatedById })

    const response = await fetch(`${this.baseUrl}/${meritId}?${params}`, {
      method: 'PUT',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(meritData),
    })

    if (!response.ok) {
      const error = await response.json()
      throw new Error(error.detail || `Failed to update merit: ${response.statusText}`)
    }
    return response.json()
  }

  /**
   * Revoke merit (soft delete)
   */
  async revokeMerit(meritId: string, deletedById: string): Promise<void> {
    const params = new URLSearchParams({ deleted_by_id: deletedById })

    const response = await fetch(`${this.baseUrl}/${meritId}?${params}`, {
      method: 'DELETE',
    })

    if (!response.ok) {
      throw new Error(`Failed to revoke merit: ${response.statusText}`)
    }
  }

  /**
   * Get student merit history
   */
  async getStudentMerits(
    studentId: string,
    filters?: { category?: string; quarter?: string; page?: number; limit?: number }
  ): Promise<MeritListResponse> {
    const params = new URLSearchParams({
      page: String(filters?.page || 1),
      limit: String(filters?.limit || 50),
    })

    if (filters?.category) {
      params.append('category', filters.category)
    }
    if (filters?.quarter) {
      params.append('quarter', filters.quarter)
    }

    const response = await fetch(`${this.baseUrl}/student/${studentId}?${params}`)
    if (!response.ok) {
      throw new Error(`Failed to fetch student merits: ${response.statusText}`)
    }
    return response.json()
  }

  /**
   * Get student merit summary
   */
  async getStudentSummary(studentId: string): Promise<MeritSummary> {
    const response = await fetch(`${this.baseUrl}/student/${studentId}/summary`)
    if (!response.ok) {
      throw new Error(`Failed to fetch student summary: ${response.statusText}`)
    }
    return response.json()
  }

  /**
   * Get class merits
   */
  async getClassMerits(
    classId: string,
    filters?: { quarter?: string; category?: string; page?: number; limit?: number }
  ): Promise<MeritListResponse> {
    const params = new URLSearchParams({
      page: String(filters?.page || 1),
      limit: String(filters?.limit || 50),
    })

    if (filters?.quarter) {
      params.append('quarter', filters.quarter)
    }
    if (filters?.category) {
      params.append('category', filters.category)
    }

    const response = await fetch(`${this.baseUrl}/class/${classId}?${params}`)
    if (!response.ok) {
      throw new Error(`Failed to fetch class merits: ${response.statusText}`)
    }
    return response.json()
  }

  /**
   * Get class merit summary
   */
  async getClassSummary(classId: string, quarter?: string): Promise<ClassMeritSummary> {
    const params = new URLSearchParams()
    if (quarter) {
      params.append('quarter', quarter)
    }

    const response = await fetch(`${this.baseUrl}/class/${classId}/summary?${params}`)
    if (!response.ok) {
      throw new Error(`Failed to fetch class summary: ${response.statusText}`)
    }
    return response.json()
  }

  /**
   * Get merits awarded by teacher
   */
  async getTeacherMerits(
    teacherId: string,
    filters?: { quarter?: string; page?: number; limit?: number }
  ): Promise<MeritListResponse> {
    const params = new URLSearchParams({
      page: String(filters?.page || 1),
      limit: String(filters?.limit || 50),
    })

    if (filters?.quarter) {
      params.append('quarter', filters.quarter)
    }

    const response = await fetch(`${this.baseUrl}/teacher/${teacherId}?${params}`)
    if (!response.ok) {
      throw new Error(`Failed to fetch teacher merits: ${response.statusText}`)
    }
    return response.json()
  }

  /**
   * Get merit leaderboard
   */
  async getLeaderboard(
    schoolId: string,
    gradeLevel?: number,
    quarter?: string,
    limit: number = 20
  ): Promise<LeaderboardEntry[]> {
    const params = new URLSearchParams({
      school_id: schoolId,
      limit: String(limit),
    })

    if (gradeLevel) {
      params.append('grade_level', String(gradeLevel))
    }
    if (quarter) {
      params.append('quarter', quarter)
    }

    const response = await fetch(`${this.baseUrl}/leaderboard/rankings?${params}`)
    if (!response.ok) {
      throw new Error(`Failed to fetch leaderboard: ${response.statusText}`)
    }
    return response.json()
  }

  /**
   * Get merit statistics
   */
  async getStatistics(
    schoolId: string,
    quarter?: string,
    gradeLevel?: number
  ): Promise<MeritStatistics> {
    const params = new URLSearchParams({ school_id: schoolId })

    if (quarter) {
      params.append('quarter', quarter)
    }
    if (gradeLevel) {
      params.append('grade_level', String(gradeLevel))
    }

    const response = await fetch(`${this.baseUrl}/statistics/summary?${params}`)
    if (!response.ok) {
      throw new Error(`Failed to fetch statistics: ${response.statusText}`)
    }
    return response.json()
  }
}

export const meritService = new MeritService()
